src/poetry/core/packages/directory_dependency.py

```diff
@@ -55,10 +40,44 @@ def __init__(
             source_url=self._full_path.as_posix(),
             extras=extras,
         )
-
         # cache this function to avoid multiple IO reads and parsing
         self.supports_poetry = functools.lru_cache(maxsize=1)(self._supports_poetry)
 
+    def _validate(self) -> None:
+        # If the directory exists do some general validation on it.
+        # There is no valid reason to have a path dependency to something
+        # that is not a valid Python project.
+        # If the directory doesn't exist, emit a warning and continue.
+        # The most common cause of a directory not existing is likely user error
+        # (e.g. a typo) so we want to do _something_ to alert them.
+        # But there are also valid situations where we might want to build
+        # a directory dependency pointing to a non-existing folder, for example:
+        # - We are re-locking a project where a path dependency was removed
+        #   It still exists in the lockfile and once we finish re-locking
+        #   we'll delete it from the lockfile, but we need to load it from the
+        #   lockfile first to begin solving.
+        # - A user wants to dynamically link to / insert the dependency at runtime.
+        #   This may seem a bit strange but in theory is totally valid and we should
+        #   allow it.
+        if not self._path.is_absolute():
+            try:
+                self._full_path = self._base.joinpath(self._path).resolve()
+            except FileNotFoundError:
+                warn(f"Directory {self._path} does not exist", category=UserWarning)
```

Comment:
I like mentioning the name of the dependency (how you did in the companion PR). Maybe, we can add it in the warning/error messages here, too.